lab_3/main.py

```diff
@@ -134,60 +245,120 @@ def get_n_grams_frequencies(self) -> int:
             (1, 5): 2, (5, 2): 2, (2, 1): 2, (1, 3): 1
         }
         """
-        pass
+        if not self.n_grams:
+            return 1
+        for i in self.n_grams:
+            for n_gramm in i:
+                for element in n_gramm:
+                    if element in self.n_gram_frequencies:
+                        self.n_gram_frequencies[element] += 1
+                    else:
+                        self.n_gram_frequencies[element] = 1
+        return 0
 
     # 8
     def extract_n_grams_frequencies(self, n_grams_dictionary: dict) -> int:
         """
         Extracts n_grams frequencies from given dictionary.
         Fills self.n_gram_frequency field.
         """
-        pass
+        if not isinstance(n_grams_dictionary, dict):
+            return 1
+        for i in n_grams_dictionary.items():
+            if isinstance(i[0], tuple) and isinstance(i[1], int):
+                self.n_gram_frequencies[i[0]] = i[1]
+        return 0
 
     # 10
     def extract_n_grams_log_probabilities(self, n_grams_dictionary: dict) -> int:
         """
         Extracts n_grams log-probabilities from given dictionary.
         Fills self.n_gram_log_probabilities field.
         """
-        pass
+        if not isinstance(n_grams_dictionary, dict):
+            return 1
+        for i in n_grams_dictionary.items():
+            if isinstance(i[0], tuple) and isinstance(i[1], float):
+                self.n_gram_log_probabilities[i[0]] = i[1]
+        return 0
 
     # 10
     def calculate_log_probabilities(self) -> int:
         """
         Gets log-probabilities of n-grams, fills the field n_gram_log_probabilities
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not self.n_gram_frequencies:
+            return 1
+
+        for n_gram in self.n_gram_frequencies:
+            probability = self.n_gram_frequencies[n_gram] / \
+                          sum([self.n_gram_frequencies[other_n_gram]
+                               for other_n_gram in self.n_gram_frequencies
+                               if n_gram[:self.size - 1] == other_n_gram[:self.size - 1]])
+            self.n_gram_log_probabilities[n_gram] = log(probability)
+        return 0
+
+
+def freq_dict_decoder(profile: NGramTrie, freq_dict: dict) -> dict:
```

Comment:
profile is not NGramTrie class object - profile is LanguageProfile class object -> bad naming leading to misunderstanding